crates/compiler/can/src/operator.rs

```diff
@@ -430,6 +479,88 @@ fn desugar_field<'a>(
     }
 }
 
+struct RecordBuilderArg<'a> {
+    closure: &'a Loc<Expr<'a>>,
+    apply_exprs: Vec<'a, &'a Loc<Expr<'a>>>,
+}
+
+fn record_builder_arg<'a>(
+    arena: &'a Bump,
+    region: Region,
+    fields: Collection<'a, Loc<RecordBuilderField<'a>>>,
+) -> RecordBuilderArg<'a> {
+    let mut record_fields = Vec::with_capacity_in(fields.len(), arena);
+    let mut apply_exprs = Vec::with_capacity_in(fields.len(), arena);
+    let mut apply_field_names = Vec::with_capacity_in(fields.len(), arena);
```

Comment:
These can end up with more capacity than necessary if the builder contains non-`ApplyValue` fields. Although, in practice most fields in a builder should be of that kind, so I think it makes sense to optimize for that case.